substrate/utils/wasm-builder/src/lib.rs

```diff
@@ -409,50 +378,62 @@ fn get_bool_environment_variable(name: &str) -> Option<bool> {
 	}
 }
 
-/// Returns whether we need to also compile the standard library when compiling the runtime.
-fn build_std_required() -> bool {
-	let default = runtime_target() == RuntimeTarget::Wasm;
-
-	crate::get_bool_environment_variable(crate::WASM_BUILD_STD).unwrap_or(default)
-}
-
 #[derive(Copy, Clone, PartialEq, Eq)]
 enum RuntimeTarget {
 	Wasm,
 	Riscv,
 }
 
 impl RuntimeTarget {
-	fn rustc_target(self) -> &'static str {
+	/// Creates a new instance.
+	fn new() -> Self {
+		let Some(value) = env::var_os(RUNTIME_TARGET) else {
+			return Self::Wasm;
+		};
+
+		if value == "wasm" {
+			Self::Wasm
+		} else if value == "riscv" {
+			Self::Riscv
+		} else {
+			build_helper::warning!(
+				"RUNTIME_TARGET environment variable must be set to either \"wasm\" or \"riscv\""
+			);
+			std::process::exit(1);
+		}
+	}
+
+	/// Figures out the target parameter value for rustc.
+	fn rustc_target(self) -> String {
 		match self {
-			RuntimeTarget::Wasm => "wasm32-unknown-unknown",
-			RuntimeTarget::Riscv => "riscv32ema-unknown-none-elf",
+			RuntimeTarget::Wasm => "wasm32-unknown-unknown".to_string(),
+			RuntimeTarget::Riscv => {
+				let path = polkavm_linker::target_json_32_path().expect("riscv not found");
+				path.into_os_string().into_string().unwrap()
+			},
 		}
 	}
 
-	fn build_subdirectory(self) -> &'static str {
-		// Keep the build directories separate so that when switching between
-		// the targets we won't trigger unnecessary rebuilds.
+	/// Figures out the target directory name used by cargo.
+	fn rustc_target_dir(self) -> &'static str {
 		match self {
-			RuntimeTarget::Wasm => "wbuild",
-			RuntimeTarget::Riscv => "rbuild",
+			RuntimeTarget::Wasm => "wasm32-unknown-unknown",
+			RuntimeTarget::Riscv => "riscv32emac-unknown-none-polkavm",
 		}
 	}
-}
 
-fn runtime_target() -> RuntimeTarget {
-	let Some(value) = env::var_os(RUNTIME_TARGET) else {
-		return RuntimeTarget::Wasm;
-	};
+	/// Figures out the build-std argument.
+	fn rustc_target_build_std(self) -> Option<&'static str> {
+		if !crate::get_bool_environment_variable(crate::WASM_BUILD_STD).unwrap_or(true) {
+			return None;
+		}
 
-	if value == "wasm" {
-		RuntimeTarget::Wasm
-	} else if value == "riscv" {
-		RuntimeTarget::Riscv
-	} else {
-		build_helper::warning!(
-			"the '{RUNTIME_TARGET}' environment variable has an invalid value; it must be either 'wasm' or 'riscv'"
-		);
-		std::process::exit(1);
+		// This is a nightly-only flag.
+		let arg = match self {
+			RuntimeTarget::Wasm => "build-std",
+			RuntimeTarget::Riscv => "build-std=core,alloc",
```

Comment:
@bkchr @jarkkojs What is the point of doing `-Zbuild-std` for target `riscv32emac-unknown-none-polkavm`? It increases runtime build time for non-wasm targets.

Reply:
It is requires for both targets but for different reasons:

wasm: The rustup provided standard library is compiled with wasm extensions enabled we dont support. Hence we need to recompile it.

riscv: Rustup doesn't contain any prebuilt binaries for this target. Hence we need to build it ourselves.

Reply:
Yeah, so I guessed, i.e. it should be official target in rustc upstream to have prebuilt binaries, right?

Reply:
[not actually at work yet, my keyboard just writes by itself]

Reply:
@athei Thanks for the clarification! I think Parity could add to the pre-compiled libraries to `riscv32emac-unknown-none-polkavm` target in `rlib` format like the `wasm32-unknown-unknown` target does (located in `./lib/rustlib/wasm32-unknown-unknown/lib/libcore-192e883b194683b6.rlib`).